### PR TITLE
[JBTM-3368] provide error messages in the HTTP responses

### PR DESCRIPTION
**Pull request by ochaloup**
https://issues.redhat.com/browse/JBTM-3368

Idea of this PR is having information on error response. This helps in debugging and test investigation and expecting for users as well.

!MAIN !CORE !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO
LRA

### CODE REVIEW

File: rts/lra/narayana-lra/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java [hunk 1]
```diff
@@ -331,6 +331,7 @@ public void filter(ContainerRequestContext containerRequestContext) {
                             toURI(terminateURIs.get(STATUS)),
                             null);
                 } catch (NotFoundException e) {
+                    lraTrace(containerRequestContext, lraId, "ServerLRAFilter before: not found with " + e.getMessage());
```
Comment (xstefank) [inline rts/lra/narayana-lra/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java:334]:
```suggestion
                    lraTrace(containerRequestContext, lraId, "ServerLRAFilter before: not found exception with " + e.getMessage());
```

seems a little out of context without this word

File: rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java [hunk 2]
```diff
@@ -739,16 +745,18 @@ private void endLRA(URI lra, boolean confirm) throws WebApplicationException {
...
                 throwGenericLRAException(lra, INTERNAL_SERVER_ERROR.getStatusCode(),
                         "LRA finished with an unexpected status code: " + response.getStatus());
             }
 
-            if (response.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
-                LRALogger.logger.infof("Could not %s LRA '%s': coordinator '%s' responded with status '%s'",
-                        confirm ? "close" : "compensate", lra, base, Response.Status.NOT_FOUND.getReasonPhrase());
-                throw new NotFoundException(lra.toASCIIString());
+            if (response.getStatus() == NOT_FOUND.getStatusCode()) {
+                String errorMsg = String.format("Could not %s LRA '%s': coordinator '%s' responded with status '%s'",
```
Comment (mmusgrov) [inline rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:755]: If you use `LRALogger.logger.infof` then you can avoid this string parsing if the level is not info.

**Review by mmusgrov — APPROVED** [review]
At some point we should discuss including error codes (a la lraI18NLogger) in the entity bodies for error responses (in many places we are using English text but having a code would enable readers to lookup the error - something like https://docs.wildfly.org/20/wildscribe/log-message-reference.html).

File: rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java [hunk 3]
```diff
@@ -706,11 +709,14 @@ private URI enlistCompensator(URI uri, long timelimit, String linkHeader, String
...
-            throw new WebApplicationException(lraId + ": Too late to join with this LRA ", PRECONDITION_FAILED);
-        } else if (response.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
-            LRALogger.logger.infof("Failed enlisting to LRA '%s', coordinator '%s' responded with status '%s'",
-                    lraId, base, Response.Status.NOT_FOUND.getStatusCode());
-            throw new WebApplicationException(uri.toASCIIString(), Response.Status.GONE);
+                String errorMsg = lraId + ": Too late to join with this LRA ";
+            throw new WebApplicationException(errorMsg,
+                    Response.status(PRECONDITION_FAILED).entity(errorMsg).build());
+        } else if (response.getStatus() == NOT_FOUND.getStatusCode()) {
+            LRALogger.logger.infof("Failed enlisting to LRA '%s', coordinator '%s' responded with status '%d (%s)'. Returning '%d (%s)'.",
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java]: we normally used `WARN` and `ERROR` levels to be part of the i18NLogger messages but I assume this is a good idea to use it here as well

File: rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java [hunk 4]
```diff
@@ -739,16 +745,18 @@ private void endLRA(URI lra, boolean confirm) throws WebApplicationException {
...
             }
 
-            if (response.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
-                LRALogger.logger.infof("Could not %s LRA '%s': coordinator '%s' responded with status '%s'",
-                        confirm ? "close" : "compensate", lra, base, Response.Status.NOT_FOUND.getReasonPhrase());
-                throw new NotFoundException(lra.toASCIIString());
+            if (response.getStatus() == NOT_FOUND.getStatusCode()) {
+                String errorMsg = String.format("Could not %s LRA '%s': coordinator '%s' responded with status '%s'",
+                        confirm ? "close" : "compensate", lra, base, NOT_FOUND.getReasonPhrase());
+                LRALogger.logger.info(errorMsg);
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:757]: +1

File: rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java [hunk 5]
```diff
@@ -289,9 +290,10 @@ public Response getNestedLRAStatus(@PathParam("NestedLraId")String nestedLraId)
 
         if (status == null || lra.getLRAStatus() == null) {
             LRALogger.i18NLogger.error_cannotGetStatusOfNestedLraURI(nestedLraId, lra.getId());
-            throw new WebApplicationException(Response.status(Response.Status.PRECONDITION_FAILED)
-                    .entity(String.format("LRA is in the wrong state for operation '%s': %s",
-                            "getNestedLRAStatus", "The LRA is still  active")).build());
+            String errMsg = String.format("LRA (parent: %s, nested:%s) is in the wrong state for operation " +
+                    "'getNestedLRAStatus: The LRA is still active", nestedLraId, lra.getId());
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java]: +1

File: rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java [hunk 6]
```diff
@@ -252,7 +256,9 @@ public synchronized URI startLRA(String baseUri, URI parentLRA, String clientId,
 
             lra.abort();
 
-            throw new InternalServerErrorException("Could not start LRA: " + ActionStatus.stringForm(status));
+            String errorMsg = "Could not start LRA: " + ActionStatus.stringForm(status);
+            throw new InternalServerErrorException(errorMsg,
+                    Response.status(INTERNAL_SERVER_ERROR).entity(errorMsg).build());
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java:261]: That's how I understand it. I was checking if I can do this better but I haven't found it. Any WebApplicationException takes the `Response` as parameter in constructor and run validity checker where it checks if the status of the response is the same as it's expected by the name of the exception. I can't find another way now.

**Comment by ochaloup** [comment]
@mmusgrov do you think that would be good to create a new jira for having the error codes as part of the error responses?
@xstefank I tried to address the points.

**Pull request merged**